// @lineai/gov-deals/build/module/datasources/factory.js
/**
 * Factory for creating data source instances
 */
import { SamApi } from '../clients/sam';
import { CsvApi } from './csv';
/**
 * Factory function to create appropriate data source
 */
export function createGovDealsApi(config) {
    switch (config.type) {
        case 'api':
            return createApiDataSource(config);
        case 'csv':
            return createCsvDataSource(config);
        default:
            throw new Error(`Unknown data source type: ${config.type}`);
    }
}
/**
 * Create SAM.gov API data source
 */
function createApiDataSource(config) {
    if (!config.apiKey) {
        throw new Error('API key is required for SAM.gov data source');
    }
    const options = {
        apiKey: config.apiKey,
        baseUrl: config.baseUrl,
        timeout: config.timeout,
        debug: config.debug,
    };
    return new SamApi(options);
}
/**
 * Create CSV data source
 */
function createCsvDataSource(config) {
    if (!config.csvPath) {
        throw new Error('CSV path is required for CSV data source');
    }
    const options = {
        csvPath: config.csvPath,
        cacheResults: true,
    };
    return new CsvApi(options);
}
/**
 * Configuration helper for historical data mode
 */
export function createHistoricalConfig(csvPath = './scratch/FY2025_opportunities.csv') {
    return {
        type: 'csv',
        csvPath,
    };
}
/**
 * Configuration helper for production mode
 */
export function createProductionConfig(apiKey, baseUrl) {
    return {
        type: 'api',
        apiKey,
        baseUrl,
    };
}
