// @lineai/gov-deals/build/main/types/common.js
"use strict";
/**
 * Common types used across the gov-deals package
 * All types are prefixed with 'GovDeals' to avoid naming conflicts
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GovDealsApiResponseSchema = exports.GovDealsErrorResponseSchema = exports.GovDealsSortOrderSchema = exports.GovDealsPaginationParamsSchema = exports.GovDealsValueRangeSchema = exports.GovDealsDateRangeSchema = exports.GovDealsContactSchema = exports.GovDealsAgencySchema = exports.GovDealsLocationSchema = void 0;
const zod_1 = require("zod");
const constants_1 = require("../config/constants");
/**
 * Location information for government contracts
 */
exports.GovDealsLocationSchema = zod_1.z.object({
    city: zod_1.z.string().optional(),
    state: constants_1.StateCodeSchema.optional(),
    zip: zod_1.z.string().optional(),
    country: zod_1.z.string().optional(),
    addressLine1: zod_1.z.string().optional(),
    addressLine2: zod_1.z.string().optional(),
});
/**
 * Federal agency information
 */
exports.GovDealsAgencySchema = zod_1.z.object({
    name: zod_1.z.string(),
    code: zod_1.z.string().optional(),
    type: zod_1.z.string().optional(),
    parentAgency: zod_1.z.string().optional(),
});
/**
 * Government contact information
 */
exports.GovDealsContactSchema = zod_1.z.object({
    name: zod_1.z.string().optional(),
    email: zod_1.z.string().email().optional(),
    phone: zod_1.z.string().optional(),
    title: zod_1.z.string().optional(),
});
/**
 * Date range for contract searches
 */
exports.GovDealsDateRangeSchema = zod_1.z.object({
    start: zod_1.z.union([zod_1.z.date(), zod_1.z.string()]),
    end: zod_1.z.union([zod_1.z.date(), zod_1.z.string()]),
});
/**
 * Value range for contract amounts
 */
exports.GovDealsValueRangeSchema = zod_1.z.object({
    min: zod_1.z.number().optional(),
    max: zod_1.z.number().optional(),
    currency: zod_1.z.string().default('USD').optional(),
});
/**
 * Pagination parameters for API requests
 */
exports.GovDealsPaginationParamsSchema = zod_1.z.object({
    page: zod_1.z.number().int().positive().optional(),
    limit: zod_1.z.number().int().positive().max(1000).optional(),
    offset: zod_1.z.number().int().nonnegative().optional(),
});
/**
 * Sort order for results
 */
exports.GovDealsSortOrderSchema = zod_1.z.enum(['asc', 'desc']);
/**
 * API error response
 */
exports.GovDealsErrorResponseSchema = zod_1.z.object({
    code: zod_1.z.string(),
    message: zod_1.z.string(),
    details: zod_1.z.record(zod_1.z.string(), zod_1.z.unknown()).optional(),
});
/**
 * API response wrapper
 */
const GovDealsApiResponseSchema = (dataSchema) => zod_1.z.object({
    data: dataSchema,
    metadata: zod_1.z.object({
        totalRecords: zod_1.z.number().optional(),
        page: zod_1.z.number().optional(),
        limit: zod_1.z.number().optional(),
        hasMore: zod_1.z.boolean().optional(),
    }).optional(),
});
exports.GovDealsApiResponseSchema = GovDealsApiResponseSchema;
