// @lineai/gov-deals/build/main/types/opportunities.js
"use strict";
/**
 * SAM.gov Opportunity types with Zod validation
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SamOpportunitySearchResponseSchema = exports.SamOpportunitySortOptionsSchema = exports.SamOpportunitySearchFiltersSchema = exports.SamOpportunitySchema = exports.SamLinkSchema = exports.SamAwardSchema = exports.SamPlaceOfPerformanceSchema = exports.SamOfficeAddressSchema = exports.SamPointOfContactSchema = exports.SamResourceLinkSchema = exports.SamAwardInfoSchema = exports.SamOpportunityDatesSchema = exports.SamAttachmentSchema = exports.SamClassificationCodeSchema = void 0;
const zod_1 = require("zod");
const constants_1 = require("../config/constants");
const common_1 = require("./common");
/**
 * Classification code (NAICS, PSC, etc.)
 */
exports.SamClassificationCodeSchema = zod_1.z.object({
    type: zod_1.z.enum(['NAICS', 'PSC', 'FSC']),
    code: zod_1.z.string(),
    description: zod_1.z.string().optional(),
});
/**
 * Attachment/document information
 */
exports.SamAttachmentSchema = zod_1.z.object({
    name: zod_1.z.string(),
    type: zod_1.z.string().optional(),
    size: zod_1.z.number().optional(),
    url: zod_1.z.string().url().optional(),
    postedDate: zod_1.z.string().optional(),
});
/**
 * Important dates for an opportunity
 */
exports.SamOpportunityDatesSchema = zod_1.z.object({
    posted: zod_1.z.string().optional(),
    modified: zod_1.z.string().optional(),
    responseDeadline: zod_1.z.string().optional(),
    archiveDate: zod_1.z.string().optional(),
    closeDate: zod_1.z.string().optional(),
});
/**
 * Award information
 */
exports.SamAwardInfoSchema = zod_1.z.object({
    awardDate: zod_1.z.string().optional(),
    awardNumber: zod_1.z.string().optional(),
    awardAmount: zod_1.z.number().optional(),
    awardee: zod_1.z.object({
        name: zod_1.z.string().optional(),
        uei: zod_1.z.string().optional(), // Unique Entity Identifier
        cageCode: zod_1.z.string().optional(),
    }).optional(),
});
/**
 * Resource link
 */
exports.SamResourceLinkSchema = zod_1.z.object({
    url: zod_1.z.string().url(),
    description: zod_1.z.string().optional(),
});
/**
 * Point of Contact schema (actual API structure)
 */
exports.SamPointOfContactSchema = zod_1.z.object({
    type: zod_1.z.string(), // "primary", "secondary", etc.
    fullName: zod_1.z.string().nullable().optional(),
    title: zod_1.z.string().nullable().optional(),
    email: zod_1.z.string().nullable().optional(), // Relaxed validation - SAM.gov has data quality issues
    phone: zod_1.z.string().nullable().optional(),
    fax: zod_1.z.string().nullable().optional(),
});
/**
 * Office Address schema (actual API structure)
 */
exports.SamOfficeAddressSchema = zod_1.z.object({
    zipcode: zod_1.z.string().nullable().optional(),
    city: zod_1.z.string().nullable().optional(),
    state: zod_1.z.string().nullable().optional(), // Allow international states
    countryCode: zod_1.z.string().nullable().optional(),
});
/**
 * Place of Performance schema (actual API structure with nested objects)
 */
exports.SamPlaceOfPerformanceSchema = zod_1.z.object({
    city: zod_1.z.object({
        code: zod_1.z.string().optional(),
        name: zod_1.z.string().optional(),
    }).nullable().optional(),
    state: zod_1.z.object({
        code: zod_1.z.string().optional(),
        name: zod_1.z.string().optional(),
    }).nullable().optional(),
    country: zod_1.z.object({
        code: zod_1.z.string().optional(),
        name: zod_1.z.string().optional(),
    }).nullable().optional(),
});
/**
 * Award schema (actual API structure)
 */
exports.SamAwardSchema = zod_1.z.object({
    awardee: zod_1.z.object({
        manual: zod_1.z.boolean().optional(),
        name: zod_1.z.string().nullable().optional(),
        uei: zod_1.z.string().nullable().optional(),
        cageCode: zod_1.z.string().nullable().optional(),
    }).optional(),
});
/**
 * Link schema
 */
exports.SamLinkSchema = zod_1.z.object({
    rel: zod_1.z.string(),
    href: zod_1.z.string(),
});
/**
 * Full SAM.gov Opportunity schema (based on actual API response)
 */
exports.SamOpportunitySchema = zod_1.z.object({
    // Core identifiers
    noticeId: zod_1.z.string(),
    title: zod_1.z.string(),
    solicitationNumber: zod_1.z.string().nullable().optional(),
    // Agency information
    fullParentPathName: zod_1.z.string().optional(),
    fullParentPathCode: zod_1.z.string().optional(),
    organizationType: zod_1.z.string().optional(),
    // Dates
    postedDate: zod_1.z.string(),
    responseDeadLine: zod_1.z.string().nullable().optional(),
    archiveDate: zod_1.z.string().optional(),
    // Type information
    type: zod_1.z.string(), // Keep as string since API returns "Solicitation" not codes
    baseType: zod_1.z.string().optional(),
    archiveType: zod_1.z.string().optional(),
    // Set-aside information
    typeOfSetAside: constants_1.TypeOfSetAsideSchema.nullable().optional().or(zod_1.z.literal('')),
    typeOfSetAsideDescription: zod_1.z.string().nullable().optional(),
    // Classification
    naicsCode: zod_1.z.string().nullable().optional(),
    naicsCodes: zod_1.z.array(zod_1.z.string()).optional(),
    classificationCode: zod_1.z.string().nullable().optional(),
    // Status
    active: constants_1.ActiveStatusSchema.optional(), // "Yes"/"No" validation
    // Description
    description: zod_1.z.string().optional(), // URL to description endpoint
    // Contacts
    pointOfContact: zod_1.z.array(exports.SamPointOfContactSchema).optional(),
    // Addresses
    officeAddress: exports.SamOfficeAddressSchema.nullable().optional(),
    placeOfPerformance: exports.SamPlaceOfPerformanceSchema.nullable().optional(),
    // Award information
    award: exports.SamAwardSchema.nullable().optional(),
    // Links
    uiLink: zod_1.z.string().optional(),
    links: zod_1.z.array(exports.SamLinkSchema).optional(),
    additionalInfoLink: zod_1.z.string().nullable().optional(),
    resourceLinks: zod_1.z.array(zod_1.z.string()).nullable().optional(), // Array of resource URLs or null
});
/**
 * Search filters for opportunities
 */
exports.SamOpportunitySearchFiltersSchema = zod_1.z.object({
    // Keywords
    keywords: zod_1.z.string().optional(),
    // Type filters
    types: zod_1.z.array(constants_1.OpportunityTypeSchema).optional(),
    // Classification filters
    naicsCodes: zod_1.z.array(zod_1.z.string()).optional(),
    pscCodes: zod_1.z.array(zod_1.z.string()).optional(),
    // Location filters
    placeOfPerformanceStates: zod_1.z.array(constants_1.StateCodeSchema).optional(),
    placeOfPerformanceZips: zod_1.z.array(zod_1.z.string()).optional(),
    // Agency filters
    agencies: zod_1.z.array(zod_1.z.string()).optional(),
    // Set-aside filters
    setAsideTypes: zod_1.z.array(constants_1.TypeOfSetAsideSchema).optional(),
    // Date filters
    postedFrom: zod_1.z.string().optional(),
    postedTo: zod_1.z.string().optional(),
    responseDeadlineFrom: zod_1.z.string().optional(),
    responseDeadlineTo: zod_1.z.string().optional(),
    // Value filters
    estimatedValueMin: zod_1.z.number().optional(),
    estimatedValueMax: zod_1.z.number().optional(),
    // Status filters
    activeOnly: zod_1.z.boolean().optional(),
    includeArchived: zod_1.z.boolean().optional(),
    // Competition filters
    competitionStrategy: zod_1.z.array(constants_1.CompetitionStrategySchema).optional(),
});
/**
 * Sort options for opportunity searches
 */
exports.SamOpportunitySortOptionsSchema = zod_1.z.object({
    field: zod_1.z.enum(['postedDate', 'modifiedDate', 'responseDeadline', 'title', 'estimatedValue']),
    order: common_1.GovDealsSortOrderSchema,
});
/**
 * API response for opportunity search (actual API structure)
 */
exports.SamOpportunitySearchResponseSchema = zod_1.z.object({
    totalRecords: zod_1.z.number(),
    limit: zod_1.z.number(),
    offset: zod_1.z.number(),
    opportunitiesData: zod_1.z.array(exports.SamOpportunitySchema),
    links: zod_1.z.array(exports.SamLinkSchema).optional(),
    searchTips: zod_1.z.array(zod_1.z.string()).optional(),
});
