// @lineai/gov-deals/build/main/clients/sam/client.js
"use strict";
/**
 * SAM.gov API Client
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SamClient = void 0;
const axios_1 = __importDefault(require("axios"));
const errors_1 = require("../../core/errors");
/**
 * Simple SAM.gov API client
 */
class SamClient {
    constructor(options) {
        if (!options.apiKey) {
            throw new Error('SAM.gov API key is required');
        }
        this.apiKey = options.apiKey;
        this.axios = axios_1.default.create({
            baseURL: options.baseUrl || 'https://api.sam.gov/prod',
            timeout: options.timeout || 30000,
            headers: {
                'Accept': 'application/json',
            },
        });
    }
    /**
     * Make a GET request
     */
    async get(path, params) {
        try {
            // Always add API key to query parameters
            const allParams = Object.assign({ api_key: this.apiKey }, params);
            const response = await this.axios.get(path, { params: allParams });
            return response.data;
        }
        catch (error) {
            this.handleError(error);
        }
    }
    /**
     * Make a POST request
     */
    async post(path, data, params) {
        try {
            // Always add API key to query parameters
            const allParams = Object.assign({ api_key: this.apiKey }, params);
            const response = await this.axios.post(path, data, { params: allParams });
            return response.data;
        }
        catch (error) {
            this.handleError(error);
        }
    }
    /**
     * Handle errors
     */
    handleError(error) {
        var _a, _b, _c, _d;
        if (axios_1.default.isAxiosError(error)) {
            const status = (_a = error.response) === null || _a === void 0 ? void 0 : _a.status;
            const message = ((_c = (_b = error.response) === null || _b === void 0 ? void 0 : _b.data) === null || _c === void 0 ? void 0 : _c.message) || error.message;
            if (status === 401 || status === 403) {
                throw new errors_1.AuthenticationError(`Authentication failed: ${message}`);
            }
            throw new errors_1.ApiError(`SAM.gov API error: ${message}`, status, (_d = error.response) === null || _d === void 0 ? void 0 : _d.data);
        }
        throw error;
    }
}
exports.SamClient = SamClient;
