// @lineai/gov-deals/build/main/clients/sam/index.js
"use strict";
/**
 * Main SAM.gov client
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SamApi = void 0;
const client_1 = require("./client");
const opportunities_1 = require("./endpoints/opportunities");
const config_1 = require("../../core/config");
/**
 * Main SAM.gov API client with all endpoints
 */
class SamApi {
    constructor(options) {
        // Create simple client options
        const clientOptions = {
            apiKey: options.apiKey,
            baseUrl: options.baseUrl,
            timeout: options.timeout,
        };
        this.client = new client_1.SamClient(clientOptions);
        this.opportunities = new opportunities_1.OpportunitiesEndpoint(this.client);
    }
    /**
     * Create SamApi from configuration
     */
    static fromConfig(config) {
        (0, config_1.validateConfig)(config);
        return new SamApi({
            apiKey: config.samGov.apiKey,
            baseUrl: config.samGov.baseUrl,
            timeout: config.samGov.timeout,
            debug: config.debug,
        });
    }
    /**
     * Create SamApi from environment variables
     */
    static fromEnv() {
        const config = (0, config_1.createConfig)();
        return SamApi.fromConfig(config);
    }
}
exports.SamApi = SamApi;
// Re-export types
__exportStar(require("./client"), exports);
__exportStar(require("./endpoints/opportunities"), exports);
__exportStar(require("../../types/opportunities"), exports);
__exportStar(require("../../config/constants"), exports);
