// @lineai/gov-deals/build/main/config/constants.js
"use strict";
/**
 * Configuration constants for the gov-deals package
 * These values can be updated as the SAM.gov API evolves
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActiveStatusSchema = exports.ACTIVE_STATUS = exports.CompetitionStrategySchema = exports.COMPETITION_STRATEGIES = exports.StateCodeSchema = exports.US_STATES = exports.API_CONFIG = exports.ContractPricingTypeSchema = exports.CONTRACT_PRICING_TYPES = exports.ConstructionNAICSSchema = exports.CONSTRUCTION_NAICS_CODES = exports.SetAsideTypeSchema = exports.TypeOfSetAsideSchema = exports.SET_ASIDE_TYPES = exports.OpportunityTypeSchema = exports.OPPORTUNITY_TYPES = void 0;
exports.getStateCode = getStateCode;
exports.isValidStateCode = isValidStateCode;
const zod_1 = require("zod");
/**
 * SAM.gov Opportunity Type Codes
 * Source: SAM.gov API Documentation
 * Last verified: 2024
 */
exports.OPPORTUNITY_TYPES = {
    PRESOLICITATION: 'p',
    SOLICITATION: 'o',
    AWARD_NOTICE: 'a',
    INTENT_TO_BUNDLE: 'i',
    JUSTIFICATION: 'u',
    SOURCES_SOUGHT: 'r',
    SPECIAL_NOTICE: 's',
    COMBINED_SYNOPSIS: 'k',
    SALE_OF_SURPLUS: 'g',
};
// Zod schema for runtime validation
exports.OpportunityTypeSchema = zod_1.z.enum([
    exports.OPPORTUNITY_TYPES.PRESOLICITATION,
    exports.OPPORTUNITY_TYPES.SOLICITATION,
    exports.OPPORTUNITY_TYPES.AWARD_NOTICE,
    exports.OPPORTUNITY_TYPES.INTENT_TO_BUNDLE,
    exports.OPPORTUNITY_TYPES.JUSTIFICATION,
    exports.OPPORTUNITY_TYPES.SOURCES_SOUGHT,
    exports.OPPORTUNITY_TYPES.SPECIAL_NOTICE,
    exports.OPPORTUNITY_TYPES.COMBINED_SYNOPSIS,
    exports.OPPORTUNITY_TYPES.SALE_OF_SURPLUS,
]);
/**
 * Federal Set-Aside Type Codes
 * Source: Federal Acquisition Regulation (FAR)
 */
exports.SET_ASIDE_TYPES = {
    SMALL_BUSINESS: 'SBA',
    SMALL_BUSINESS_SET_ASIDE: 'SBP',
    EIGHT_A: '8A',
    EIGHT_A_SOLE_SOURCE: '8AN',
    EIGHT_A_COMPETITIVE: '8AC',
    HUB_ZONE: 'HZC',
    HUB_ZONE_SOLE_SOURCE: 'HZS',
    SDVOSB: 'SDVOSBC',
    SDVOSB_SOLE_SOURCE: 'SDVOSBS',
    WOMEN_OWNED: 'WOSB',
    EDWOSB: 'EDWOSB',
    LOCAL_AREA: 'LAS',
    INDIAN_ECONOMIC: 'IEE',
    INDIAN_SMALL_BUSINESS: 'ISBEE',
    BUY_INDIAN: 'BICiv',
    VETERAN_OWNED: 'VSA',
    VETERAN_OWNED_SOLE_SOURCE: 'VSS',
    NONE: 'NONE',
};
// Renamed to match actual API field name
exports.TypeOfSetAsideSchema = zod_1.z.enum([
    exports.SET_ASIDE_TYPES.SMALL_BUSINESS,
    exports.SET_ASIDE_TYPES.SMALL_BUSINESS_SET_ASIDE,
    exports.SET_ASIDE_TYPES.EIGHT_A,
    exports.SET_ASIDE_TYPES.EIGHT_A_SOLE_SOURCE,
    exports.SET_ASIDE_TYPES.EIGHT_A_COMPETITIVE,
    exports.SET_ASIDE_TYPES.HUB_ZONE,
    exports.SET_ASIDE_TYPES.HUB_ZONE_SOLE_SOURCE,
    exports.SET_ASIDE_TYPES.SDVOSB,
    exports.SET_ASIDE_TYPES.SDVOSB_SOLE_SOURCE,
    exports.SET_ASIDE_TYPES.WOMEN_OWNED,
    exports.SET_ASIDE_TYPES.EDWOSB,
    exports.SET_ASIDE_TYPES.LOCAL_AREA,
    exports.SET_ASIDE_TYPES.INDIAN_ECONOMIC,
    exports.SET_ASIDE_TYPES.INDIAN_SMALL_BUSINESS,
    exports.SET_ASIDE_TYPES.BUY_INDIAN,
    exports.SET_ASIDE_TYPES.VETERAN_OWNED,
    exports.SET_ASIDE_TYPES.VETERAN_OWNED_SOLE_SOURCE,
    exports.SET_ASIDE_TYPES.NONE,
]);
// Keep the old name for backward compatibility
exports.SetAsideTypeSchema = exports.TypeOfSetAsideSchema;
/**
 * Construction-related NAICS Codes
 * Source: North American Industry Classification System
 */
exports.CONSTRUCTION_NAICS_CODES = {
    COMMERCIAL_BUILDING: '236220',
    INDUSTRIAL_BUILDING: '236210',
    RESIDENTIAL_REMODELING: '236118',
    ELECTRICAL_CONTRACTORS: '238210',
    PLUMBING_HVAC: '238220',
    DRYWALL_INSULATION: '238310',
    PAINTING: '238320',
    SITE_PREPARATION: '238910',
    ROOFING: '238160',
    CONCRETE: '238110',
};
// Create an array of values for the schema
const constructionNAICSValues = Object.values(exports.CONSTRUCTION_NAICS_CODES);
exports.ConstructionNAICSSchema = zod_1.z.enum(constructionNAICSValues);
/**
 * Contract Pricing Types
 * Source: Federal Acquisition Regulation (FAR) Part 16
 */
exports.CONTRACT_PRICING_TYPES = {
    FIRM_FIXED_PRICE: 'FFP',
    FIXED_PRICE_EPA: 'FPE',
    FIXED_PRICE_INCENTIVE: 'FPI',
    FIXED_PRICE_REDETERMINATION: 'FPR',
    COST_PLUS_FIXED_FEE: 'CPFF',
    COST_PLUS_INCENTIVE_FEE: 'CPIF',
    COST_PLUS_AWARD_FEE: 'CPAF',
    COST_SHARING: 'CS',
    COST_REIMBURSEMENT: 'CR',
    TIME_AND_MATERIALS: 'TM',
    LABOR_HOURS: 'LH',
    OTHER: 'OT',
};
const contractPricingValues = Object.values(exports.CONTRACT_PRICING_TYPES);
exports.ContractPricingTypeSchema = zod_1.z.enum(contractPricingValues);
/**
 * API Configuration
 */
exports.API_CONFIG = {
    SAM_GOV: {
        BASE_URL: 'https://api.sam.gov/prod',
        OPPORTUNITIES_ENDPOINT: '/opportunities/v2/search',
        ENTITIES_ENDPOINT: '/entity-information/v3/entities',
        HIERARCHY_ENDPOINT: '/federalorganizations/v2/hierarchy',
        DESCRIPTION_ENDPOINT: '/opportunities/v1/noticedesc',
        DEFAULT_PAGE_SIZE: 100,
        MAX_PAGE_SIZE: 1000,
    },
    RETRY: {
        MAX_ATTEMPTS: 3,
        INITIAL_DELAY_MS: 1000,
        MAX_DELAY_MS: 10000,
    },
    CACHE: {
        DEFAULT_TTL_SECONDS: 300, // 5 minutes
        MAX_CACHE_SIZE: 100,
    },
};
/**
 * US State Codes
 */
exports.US_STATES = {
    ALABAMA: 'AL',
    ALASKA: 'AK',
    ARIZONA: 'AZ',
    ARKANSAS: 'AR',
    CALIFORNIA: 'CA',
    COLORADO: 'CO',
    CONNECTICUT: 'CT',
    DELAWARE: 'DE',
    FLORIDA: 'FL',
    GEORGIA: 'GA',
    HAWAII: 'HI',
    IDAHO: 'ID',
    ILLINOIS: 'IL',
    INDIANA: 'IN',
    IOWA: 'IA',
    KANSAS: 'KS',
    KENTUCKY: 'KY',
    LOUISIANA: 'LA',
    MAINE: 'ME',
    MARYLAND: 'MD',
    MASSACHUSETTS: 'MA',
    MICHIGAN: 'MI',
    MINNESOTA: 'MN',
    MISSISSIPPI: 'MS',
    MISSOURI: 'MO',
    MONTANA: 'MT',
    NEBRASKA: 'NE',
    NEVADA: 'NV',
    NEW_HAMPSHIRE: 'NH',
    NEW_JERSEY: 'NJ',
    NEW_MEXICO: 'NM',
    NEW_YORK: 'NY',
    NORTH_CAROLINA: 'NC',
    NORTH_DAKOTA: 'ND',
    OHIO: 'OH',
    OKLAHOMA: 'OK',
    OREGON: 'OR',
    PENNSYLVANIA: 'PA',
    RHODE_ISLAND: 'RI',
    SOUTH_CAROLINA: 'SC',
    SOUTH_DAKOTA: 'SD',
    TENNESSEE: 'TN',
    TEXAS: 'TX',
    UTAH: 'UT',
    VERMONT: 'VT',
    VIRGINIA: 'VA',
    WASHINGTON: 'WA',
    WEST_VIRGINIA: 'WV',
    WISCONSIN: 'WI',
    WYOMING: 'WY',
    DISTRICT_OF_COLUMBIA: 'DC',
    PUERTO_RICO: 'PR',
    US_VIRGIN_ISLANDS: 'VI',
    GUAM: 'GU',
    AMERICAN_SAMOA: 'AS',
    NORTHERN_MARIANA_ISLANDS: 'MP',
};
const stateValues = Object.values(exports.US_STATES);
exports.StateCodeSchema = zod_1.z.enum(stateValues);
/**
 * Helper function to get state code from name
 */
function getStateCode(stateName) {
    return exports.US_STATES[stateName];
}
/**
 * Helper function to validate if a string is a valid state code
 */
function isValidStateCode(code) {
    return exports.StateCodeSchema.safeParse(code).success;
}
/**
 * Competition Strategy
 */
exports.COMPETITION_STRATEGIES = {
    FULL_AND_OPEN: 'fullAndOpen',
    SET_ASIDE: 'setAside',
    SOLE_SOURCE: 'soleSource',
    LIMITED_SOURCES: 'limitedSources',
};
const competitionValues = Object.values(exports.COMPETITION_STRATEGIES);
exports.CompetitionStrategySchema = zod_1.z.enum(competitionValues);
/**
 * Active Status for opportunities (as returned by API)
 */
exports.ACTIVE_STATUS = {
    YES: 'Yes',
    NO: 'No',
};
exports.ActiveStatusSchema = zod_1.z.enum([exports.ACTIVE_STATUS.YES, exports.ACTIVE_STATUS.NO]);
