// @lineai/gov-deals/build/main/core/config.js
"use strict";
/**
 * Configuration management for the gov-deals package
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfigSchema = void 0;
exports.createConfig = createConfig;
exports.validateConfig = validateConfig;
const zod_1 = require("zod");
const dotenv_1 = __importDefault(require("dotenv"));
// Load environment variables
dotenv_1.default.config();
/**
 * Configuration schema for validation
 */
exports.ConfigSchema = zod_1.z.object({
    samGov: zod_1.z.object({
        apiKey: zod_1.z.string().min(1, 'SAM.gov API key is required'),
        baseUrl: zod_1.z.string().url().default('https://api.sam.gov'),
        timeout: zod_1.z.number().int().positive().default(30000),
        retryAttempts: zod_1.z.number().int().nonnegative().default(3),
        retryDelay: zod_1.z.number().int().positive().default(1000),
    }),
    cache: zod_1.z.object({
        enabled: zod_1.z.boolean().default(true),
        ttlSeconds: zod_1.z.number().int().positive().default(300),
        maxSize: zod_1.z.number().int().positive().default(100),
    }),
    debug: zod_1.z.boolean().default(false),
});
/**
 * Create configuration from environment variables and options
 */
function createConfig(options = {}) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
    // Build final config with fallbacks
    const config = {
        samGov: {
            apiKey: ((_a = options.samGov) === null || _a === void 0 ? void 0 : _a.apiKey) || process.env.SAM_GOV_API_KEY || '',
            baseUrl: ((_b = options.samGov) === null || _b === void 0 ? void 0 : _b.baseUrl) || process.env.SAM_GOV_BASE_URL || 'https://api.sam.gov',
            timeout: ((_c = options.samGov) === null || _c === void 0 ? void 0 : _c.timeout) ||
                (process.env.SAM_GOV_TIMEOUT ? parseInt(process.env.SAM_GOV_TIMEOUT) : 30000),
            retryAttempts: ((_d = options.samGov) === null || _d === void 0 ? void 0 : _d.retryAttempts) ||
                (process.env.SAM_GOV_RETRY_ATTEMPTS ? parseInt(process.env.SAM_GOV_RETRY_ATTEMPTS) : 3),
            retryDelay: ((_e = options.samGov) === null || _e === void 0 ? void 0 : _e.retryDelay) ||
                (process.env.SAM_GOV_RETRY_DELAY ? parseInt(process.env.SAM_GOV_RETRY_DELAY) : 1000),
        },
        cache: {
            enabled: (_g = (_f = options.cache) === null || _f === void 0 ? void 0 : _f.enabled) !== null && _g !== void 0 ? _g : (process.env.CACHE_ENABLED ? process.env.CACHE_ENABLED === 'true' : true),
            ttlSeconds: ((_h = options.cache) === null || _h === void 0 ? void 0 : _h.ttlSeconds) ||
                (process.env.CACHE_TTL_SECONDS ? parseInt(process.env.CACHE_TTL_SECONDS) : 300),
            maxSize: ((_j = options.cache) === null || _j === void 0 ? void 0 : _j.maxSize) ||
                (process.env.CACHE_MAX_SIZE ? parseInt(process.env.CACHE_MAX_SIZE) : 100),
        },
        debug: (_k = options.debug) !== null && _k !== void 0 ? _k : ((process.env.DEBUG === 'true') || (process.env.NODE_ENV === 'development')),
    };
    // Validate the final configuration
    return exports.ConfigSchema.parse(config);
}
/**
 * Validate that required configuration is present
 */
function validateConfig(config) {
    if (!config.samGov.apiKey) {
        throw new Error('SAM.gov API key is required. Set SAM_GOV_API_KEY environment variable or pass it in configuration.');
    }
}
