// @lineai/gov-deals/build/main/core/errors.js
"use strict";
/**
 * Custom error types for the gov-deals package
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimeoutError = exports.NetworkError = exports.ValidationError = exports.RateLimitError = exports.ApiError = exports.AuthenticationError = exports.ConfigurationError = exports.GovDealsError = void 0;
exports.isGovDealsError = isGovDealsError;
exports.isRateLimitError = isRateLimitError;
exports.isRetryableError = isRetryableError;
/**
 * Base error class for all gov-deals errors
 */
class GovDealsError extends Error {
    constructor(message, code, details) {
        super(message);
        this.code = code;
        this.details = details;
        this.name = 'GovDealsError';
        // Maintains proper stack trace for where error was thrown
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, this.constructor);
        }
    }
}
exports.GovDealsError = GovDealsError;
/**
 * Configuration error
 */
class ConfigurationError extends GovDealsError {
    constructor(message, details) {
        super(message, 'CONFIGURATION_ERROR', details);
        this.name = 'ConfigurationError';
    }
}
exports.ConfigurationError = ConfigurationError;
/**
 * Authentication error for API requests
 */
class AuthenticationError extends GovDealsError {
    constructor(message, details) {
        super(message, 'AUTHENTICATION_ERROR', details);
        this.name = 'AuthenticationError';
    }
}
exports.AuthenticationError = AuthenticationError;
/**
 * API request error
 */
class ApiError extends GovDealsError {
    constructor(message, statusCode, response, details) {
        super(message, 'API_ERROR', details);
        this.statusCode = statusCode;
        this.response = response;
        this.name = 'ApiError';
    }
}
exports.ApiError = ApiError;
/**
 * Rate limit error
 */
class RateLimitError extends ApiError {
    constructor(message, retryAfter, details) {
        super(message, 429, undefined, details);
        this.retryAfter = retryAfter;
        this.name = 'RateLimitError';
    }
}
exports.RateLimitError = RateLimitError;
/**
 * Validation error for data validation failures
 */
class ValidationError extends GovDealsError {
    constructor(message, validationErrors, details) {
        super(message, 'VALIDATION_ERROR', details);
        this.validationErrors = validationErrors;
        this.name = 'ValidationError';
    }
}
exports.ValidationError = ValidationError;
/**
 * Network error for connection issues
 */
class NetworkError extends GovDealsError {
    constructor(message, originalError, details) {
        super(message, 'NETWORK_ERROR', details);
        this.originalError = originalError;
        this.name = 'NetworkError';
    }
}
exports.NetworkError = NetworkError;
/**
 * Timeout error
 */
class TimeoutError extends NetworkError {
    constructor(message, details) {
        super(message, undefined, details);
        this.name = 'TimeoutError';
    }
}
exports.TimeoutError = TimeoutError;
/**
 * Check if an error is a gov-deals error
 */
function isGovDealsError(error) {
    return error instanceof GovDealsError;
}
/**
 * Check if an error is a rate limit error
 */
function isRateLimitError(error) {
    return error instanceof RateLimitError;
}
/**
 * Check if an error is retryable
 */
function isRetryableError(error) {
    if (error instanceof NetworkError || error instanceof TimeoutError) {
        return true;
    }
    if (error instanceof ApiError) {
        // Retry on 5xx errors and specific 4xx errors
        const statusCode = error.statusCode;
        return statusCode ? statusCode >= 500 || statusCode === 429 || statusCode === 408 : false;
    }
    return false;
}
