// @lineai/gov-deals/build/main/clients/supabase-wrapper.js
"use strict";
/**
 * Wrapper to make SamApi work with Supabase Edge Functions
 * Converts query parameter authentication to header authentication
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SupabaseApi = void 0;
exports.createSupabaseApi = createSupabaseApi;
const opportunities_1 = require("./sam/endpoints/opportunities");
const axios_1 = __importDefault(require("axios"));
const errors_1 = require("../core/errors");
/**
 * Custom client that sends API key in headers instead of query params
 */
class SupabaseClient {
    constructor(apiKey, baseUrl) {
        this.axios = axios_1.default.create({
            baseURL: baseUrl,
            timeout: 30000,
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json',
                'apikey': apiKey,
                'Authorization': `Bearer ${apiKey}`,
            },
        });
        // Intercept requests to remove api_key from query params
        this.axios.interceptors.request.use((config) => {
            var _a;
            if ((_a = config.params) === null || _a === void 0 ? void 0 : _a.api_key) {
                delete config.params.api_key;
            }
            return config;
        });
    }
    async get(path, params) {
        try {
            const response = await this.axios.get(path, { params });
            return response.data;
        }
        catch (error) {
            this.handleError(error);
        }
    }
    async post(path, data, params) {
        try {
            const response = await this.axios.post(path, data, { params });
            return response.data;
        }
        catch (error) {
            this.handleError(error);
        }
    }
    handleError(error) {
        var _a, _b, _c, _d, _e, _f;
        if (axios_1.default.isAxiosError(error)) {
            const status = (_a = error.response) === null || _a === void 0 ? void 0 : _a.status;
            const message = ((_c = (_b = error.response) === null || _b === void 0 ? void 0 : _b.data) === null || _c === void 0 ? void 0 : _c.error) || ((_e = (_d = error.response) === null || _d === void 0 ? void 0 : _d.data) === null || _e === void 0 ? void 0 : _e.message) || error.message;
            if (status === 401 || status === 403) {
                throw new errors_1.AuthenticationError(`Authentication failed: ${message}`);
            }
            throw new errors_1.ApiError(`API error: ${message}`, status, (_f = error.response) === null || _f === void 0 ? void 0 : _f.data);
        }
        throw error;
    }
}
/**
 * Supabase-compatible API wrapper
 * Use this instead of SamApi when connecting to Supabase Edge Functions
 */
class SupabaseApi {
    constructor(options) {
        this.client = new SupabaseClient(options.apiKey, options.baseUrl);
        this.opportunities = new opportunities_1.OpportunitiesEndpoint(this.client);
    }
}
exports.SupabaseApi = SupabaseApi;
/**
 * Helper function to create Supabase-compatible API instance
 */
function createSupabaseApi(supabaseUrl, supabaseAnonKey) {
    // Ensure the URL includes /functions/v1/api
    const baseUrl = supabaseUrl.includes('/functions/v1')
        ? supabaseUrl
        : `${supabaseUrl}/functions/v1/api`;
    return new SupabaseApi({
        apiKey: supabaseAnonKey,
        baseUrl: baseUrl,
    });
}
