// @lineai/gov-deals/build/main/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SET_ASIDE_TYPES = exports.OPPORTUNITY_TYPES = exports.CONSTRUCTION_NAICS_CODES = exports.createConfig = exports.createProductionConfig = exports.createHistoricalConfig = exports.createGovDealsApi = exports.CsvApi = exports.createSupabaseApi = exports.SupabaseApi = exports.SamApi = void 0;
// Main package exports
var sam_1 = require("./clients/sam");
Object.defineProperty(exports, "SamApi", { enumerable: true, get: function () { return sam_1.SamApi; } });
// Supabase wrapper for header-based auth
var supabase_wrapper_1 = require("./clients/supabase-wrapper");
Object.defineProperty(exports, "SupabaseApi", { enumerable: true, get: function () { return supabase_wrapper_1.SupabaseApi; } });
Object.defineProperty(exports, "createSupabaseApi", { enumerable: true, get: function () { return supabase_wrapper_1.createSupabaseApi; } });
// Data sources
var csv_1 = require("./datasources/csv");
Object.defineProperty(exports, "CsvApi", { enumerable: true, get: function () { return csv_1.CsvApi; } });
// Factory for creating data sources
var factory_1 = require("./datasources/factory");
Object.defineProperty(exports, "createGovDealsApi", { enumerable: true, get: function () { return factory_1.createGovDealsApi; } });
Object.defineProperty(exports, "createHistoricalConfig", { enumerable: true, get: function () { return factory_1.createHistoricalConfig; } });
Object.defineProperty(exports, "createProductionConfig", { enumerable: true, get: function () { return factory_1.createProductionConfig; } });
// Configuration
var config_1 = require("./core/config");
Object.defineProperty(exports, "createConfig", { enumerable: true, get: function () { return config_1.createConfig; } });
// Types
__exportStar(require("./types/common"), exports);
__exportStar(require("./types/opportunities"), exports);
// Constants
__exportStar(require("./config/constants"), exports);
// Errors
__exportStar(require("./core/errors"), exports);
// Convenience exports for common use cases
var constants_1 = require("./config/constants");
Object.defineProperty(exports, "CONSTRUCTION_NAICS_CODES", { enumerable: true, get: function () { return constants_1.CONSTRUCTION_NAICS_CODES; } });
Object.defineProperty(exports, "OPPORTUNITY_TYPES", { enumerable: true, get: function () { return constants_1.OPPORTUNITY_TYPES; } });
Object.defineProperty(exports, "SET_ASIDE_TYPES", { enumerable: true, get: function () { return constants_1.SET_ASIDE_TYPES; } });
// NAICS utilities
__exportStar(require("./utils/naics"), exports);
