// @lineai/gov-deals/build/module/clients/sam/client.js
/**
 * SAM.gov API Client
 */
import axios from 'axios';
import { ApiError, AuthenticationError } from '../../core/errors';
/**
 * Simple SAM.gov API client
 */
export class SamClient {
    axios;
    apiKey;
    constructor(options) {
        if (!options.apiKey) {
            throw new Error('SAM.gov API key is required');
        }
        this.apiKey = options.apiKey;
        this.axios = axios.create({
            baseURL: options.baseUrl || 'https://api.sam.gov/prod',
            timeout: options.timeout || 30000,
            headers: {
                'Accept': 'application/json',
            },
        });
    }
    /**
     * Make a GET request
     */
    async get(path, params) {
        try {
            // Always add API key to query parameters
            const allParams = {
                api_key: this.apiKey,
                ...params,
            };
            const response = await this.axios.get(path, { params: allParams });
            return response.data;
        }
        catch (error) {
            this.handleError(error);
        }
    }
    /**
     * Make a POST request
     */
    async post(path, data, params) {
        try {
            // Always add API key to query parameters
            const allParams = {
                api_key: this.apiKey,
                ...params,
            };
            const response = await this.axios.post(path, data, { params: allParams });
            return response.data;
        }
        catch (error) {
            this.handleError(error);
        }
    }
    /**
     * Handle errors
     */
    handleError(error) {
        if (axios.isAxiosError(error)) {
            const status = error.response?.status;
            const message = error.response?.data?.message || error.message;
            if (status === 401 || status === 403) {
                throw new AuthenticationError(`Authentication failed: ${message}`);
            }
            throw new ApiError(`SAM.gov API error: ${message}`, status, error.response?.data);
        }
        throw error;
    }
}
