// @lineai/gov-deals/build/main/datasources/csv/client.js
"use strict";
/**
 * CSV-based data source for opportunities
 * Provides same interface as SAM.gov API but reads from local CSV file
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CsvClient = void 0;
const fs_1 = require("fs");
const sync_1 = require("csv-parse/sync");
const field_mapping_1 = require("./field-mapping");
const opportunities_1 = require("../../types/opportunities");
const errors_1 = require("../../core/errors");
/**
 * CSV-based client that mimics SAM.gov API
 */
class CsvClient {
    constructor(options) {
        var _a, _b;
        this.cache = null;
        if (!options.csvPath) {
            throw new Error('CSV file path is required');
        }
        this.csvPath = options.csvPath;
        this.cacheResults = (_a = options.cacheResults) !== null && _a !== void 0 ? _a : true;
        this.encoding = (_b = options.encoding) !== null && _b !== void 0 ? _b : 'utf-8';
    }
    /**
     * Load and parse CSV data
     */
    async loadData() {
        // Return cached data if available
        if (this.cache && this.cacheResults) {
            return this.cache;
        }
        try {
            // Read CSV file
            const csvContent = await fs_1.promises.readFile(this.csvPath, this.encoding);
            // Parse CSV with proper options
            const records = (0, sync_1.parse)(csvContent, {
                columns: true, // Use first row as column names
                skip_empty_lines: true,
                relax_quotes: true,
                relax_column_count: true,
                skip_records_with_error: true, // Skip problematic rows
            });
            // Cache if enabled
            if (this.cacheResults) {
                this.cache = records;
            }
            return records;
        }
        catch (error) {
            throw new Error(`Failed to load CSV file: ${error instanceof Error ? error.message : String(error)}`);
        }
    }
    /**
     * Search opportunities with filtering and pagination
     */
    async search(filters, pagination) {
        // Load all data
        const allRecords = await this.loadData();
        // Apply filters
        let filteredRecords = allRecords;
        if (filters) {
            filteredRecords = allRecords.filter(row => (0, field_mapping_1.filterCsvRow)(row, filters));
        }
        // Calculate pagination
        const limit = (pagination === null || pagination === void 0 ? void 0 : pagination.limit) || 100;
        const page = (pagination === null || pagination === void 0 ? void 0 : pagination.page) || 1;
        const offset = (page - 1) * limit;
        // Get paginated results
        const paginatedRecords = filteredRecords.slice(offset, offset + limit);
        // Transform to SAM format
        const opportunities = paginatedRecords.map(field_mapping_1.mapCsvToSamOpportunity);
        // Build response
        const response = {
            totalRecords: filteredRecords.length,
            limit,
            offset,
            opportunitiesData: opportunities,
            links: [], // CSV doesn't have links
        };
        // Validate response format
        const result = opportunities_1.SamOpportunitySearchResponseSchema.safeParse(response);
        if (!result.success) {
            console.error('CSV mapping validation errors:', result.error.issues);
            // Continue anyway - the data is still useful even if not perfectly typed
        }
        return response;
    }
    /**
     * Get opportunity by ID
     */
    async getById(noticeId) {
        const allRecords = await this.loadData();
        const record = allRecords.find(row => row.NoticeId === noticeId);
        if (!record) {
            throw new errors_1.ValidationError(`Opportunity with ID ${noticeId} not found`);
        }
        return (0, field_mapping_1.mapCsvToSamOpportunity)(record);
    }
    /**
     * Get opportunity description (from CSV Description field)
     */
    async getDescription(noticeId) {
        const allRecords = await this.loadData();
        const record = allRecords.find(row => row.NoticeId === noticeId);
        if (!record) {
            throw new errors_1.ValidationError(`Opportunity with ID ${noticeId} not found`);
        }
        return record.Description || 'No description available';
    }
    /**
     * Search for construction opportunities
     */
    async searchConstruction(additionalFilters, pagination) {
        const constructionFilters = Object.assign({ naicsCodes: ['236', '238'], keywords: 'renovation construction building modernization' }, additionalFilters);
        return this.search(constructionFilters, pagination);
    }
    /**
     * Clear cache if needed
     */
    clearCache() {
        this.cache = null;
    }
}
exports.CsvClient = CsvClient;
