// @lineai/gov-deals/build/main/datasources/factory.js
"use strict";
/**
 * Factory for creating data source instances
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createGovDealsApi = createGovDealsApi;
exports.createHistoricalConfig = createHistoricalConfig;
exports.createProductionConfig = createProductionConfig;
const sam_1 = require("../clients/sam");
const csv_1 = require("./csv");
/**
 * Factory function to create appropriate data source
 */
function createGovDealsApi(config) {
    switch (config.type) {
        case 'api':
            return createApiDataSource(config);
        case 'csv':
            return createCsvDataSource(config);
        default:
            throw new Error(`Unknown data source type: ${config.type}`);
    }
}
/**
 * Create SAM.gov API data source
 */
function createApiDataSource(config) {
    if (!config.apiKey) {
        throw new Error('API key is required for SAM.gov data source');
    }
    const options = {
        apiKey: config.apiKey,
        baseUrl: config.baseUrl,
        timeout: config.timeout,
        debug: config.debug,
    };
    return new sam_1.SamApi(options);
}
/**
 * Create CSV data source
 */
function createCsvDataSource(config) {
    if (!config.csvPath) {
        throw new Error('CSV path is required for CSV data source');
    }
    const options = {
        csvPath: config.csvPath,
        cacheResults: true,
    };
    return new csv_1.CsvApi(options);
}
/**
 * Configuration helper for historical data mode
 */
function createHistoricalConfig(csvPath = './scratch/FY2025_opportunities.csv') {
    return {
        type: 'csv',
        csvPath,
    };
}
/**
 * Configuration helper for production mode
 */
function createProductionConfig(apiKey, baseUrl) {
    return {
        type: 'api',
        apiKey,
        baseUrl,
    };
}
