// @lineai/gov-deals/build/main/datasources/csv/endpoints/opportunities.js
"use strict";
/**
 * CSV-based Opportunities endpoint
 * Implements the same interface as SAM.gov opportunities endpoint
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CsvOpportunitiesEndpoint = void 0;
class CsvOpportunitiesEndpoint {
    constructor(client) {
        this.client = client;
    }
    /**
     * Search for opportunities in CSV data
     */
    async search(filters, pagination) {
        return this.client.search(filters, pagination);
    }
    /**
     * Get opportunity by ID from CSV
     */
    async getById(noticeId) {
        return this.client.getById(noticeId);
    }
    /**
     * Get opportunity description from CSV
     */
    async getDescription(noticeId) {
        return this.client.getDescription(noticeId);
    }
    /**
     * Search for construction opportunities in CSV
     */
    async searchConstruction(additionalFilters, pagination) {
        return this.client.searchConstruction(additionalFilters, pagination);
    }
}
exports.CsvOpportunitiesEndpoint = CsvOpportunitiesEndpoint;
