// @lineai/gov-deals/build/module/clients/supabase-wrapper.js
/**
 * Wrapper to make SamApi work with Supabase Edge Functions
 * Converts query parameter authentication to header authentication
 */
import { OpportunitiesEndpoint } from './sam/endpoints/opportunities';
import axios from 'axios';
import { ApiError, AuthenticationError } from '../core/errors';
/**
 * Custom client that sends API key in headers instead of query params
 */
class SupabaseClient {
    axios;
    constructor(apiKey, baseUrl) {
        this.axios = axios.create({
            baseURL: baseUrl,
            timeout: 30000,
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json',
                'apikey': apiKey,
                'Authorization': `Bearer ${apiKey}`,
            },
        });
        // Intercept requests to remove api_key from query params
        this.axios.interceptors.request.use((config) => {
            if (config.params?.api_key) {
                delete config.params.api_key;
            }
            return config;
        });
    }
    async get(path, params) {
        try {
            const response = await this.axios.get(path, { params });
            return response.data;
        }
        catch (error) {
            this.handleError(error);
        }
    }
    async post(path, data, params) {
        try {
            const response = await this.axios.post(path, data, { params });
            return response.data;
        }
        catch (error) {
            this.handleError(error);
        }
    }
    handleError(error) {
        if (axios.isAxiosError(error)) {
            const status = error.response?.status;
            const message = error.response?.data?.error || error.response?.data?.message || error.message;
            if (status === 401 || status === 403) {
                throw new AuthenticationError(`Authentication failed: ${message}`);
            }
            throw new ApiError(`API error: ${message}`, status, error.response?.data);
        }
        throw error;
    }
}
/**
 * Supabase-compatible API wrapper
 * Use this instead of SamApi when connecting to Supabase Edge Functions
 */
export class SupabaseApi {
    client;
    opportunities;
    constructor(options) {
        this.client = new SupabaseClient(options.apiKey, options.baseUrl);
        this.opportunities = new OpportunitiesEndpoint(this.client);
    }
}
/**
 * Helper function to create Supabase-compatible API instance
 */
export function createSupabaseApi(supabaseUrl, supabaseAnonKey) {
    // Ensure the URL includes /functions/v1/api
    const baseUrl = supabaseUrl.includes('/functions/v1')
        ? supabaseUrl
        : `${supabaseUrl}/functions/v1/api`;
    return new SupabaseApi({
        apiKey: supabaseAnonKey,
        baseUrl: baseUrl,
    });
}
