// @lineai/gov-deals/build/module/core/errors.js
/**
 * Custom error types for the gov-deals package
 */
/**
 * Base error class for all gov-deals errors
 */
export class GovDealsError extends Error {
    code;
    details;
    constructor(message, code, details) {
        super(message);
        this.code = code;
        this.details = details;
        this.name = 'GovDealsError';
        // Maintains proper stack trace for where error was thrown
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, this.constructor);
        }
    }
}
/**
 * Configuration error
 */
export class ConfigurationError extends GovDealsError {
    constructor(message, details) {
        super(message, 'CONFIGURATION_ERROR', details);
        this.name = 'ConfigurationError';
    }
}
/**
 * Authentication error for API requests
 */
export class AuthenticationError extends GovDealsError {
    constructor(message, details) {
        super(message, 'AUTHENTICATION_ERROR', details);
        this.name = 'AuthenticationError';
    }
}
/**
 * API request error
 */
export class ApiError extends GovDealsError {
    statusCode;
    response;
    constructor(message, statusCode, response, details) {
        super(message, 'API_ERROR', details);
        this.statusCode = statusCode;
        this.response = response;
        this.name = 'ApiError';
    }
}
/**
 * Rate limit error
 */
export class RateLimitError extends ApiError {
    retryAfter;
    constructor(message, retryAfter, details) {
        super(message, 429, undefined, details);
        this.retryAfter = retryAfter;
        this.name = 'RateLimitError';
    }
}
/**
 * Validation error for data validation failures
 */
export class ValidationError extends GovDealsError {
    validationErrors;
    constructor(message, validationErrors, details) {
        super(message, 'VALIDATION_ERROR', details);
        this.validationErrors = validationErrors;
        this.name = 'ValidationError';
    }
}
/**
 * Network error for connection issues
 */
export class NetworkError extends GovDealsError {
    originalError;
    constructor(message, originalError, details) {
        super(message, 'NETWORK_ERROR', details);
        this.originalError = originalError;
        this.name = 'NetworkError';
    }
}
/**
 * Timeout error
 */
export class TimeoutError extends NetworkError {
    constructor(message, details) {
        super(message, undefined, details);
        this.name = 'TimeoutError';
    }
}
/**
 * Check if an error is a gov-deals error
 */
export function isGovDealsError(error) {
    return error instanceof GovDealsError;
}
/**
 * Check if an error is a rate limit error
 */
export function isRateLimitError(error) {
    return error instanceof RateLimitError;
}
/**
 * Check if an error is retryable
 */
export function isRetryableError(error) {
    if (error instanceof NetworkError || error instanceof TimeoutError) {
        return true;
    }
    if (error instanceof ApiError) {
        // Retry on 5xx errors and specific 4xx errors
        const statusCode = error.statusCode;
        return statusCode ? statusCode >= 500 || statusCode === 429 || statusCode === 408 : false;
    }
    return false;
}
